### ISSUE TITLE
SplitLaw doesn't hold in general

### ISSUE THREAD

**Issue by peterneyens** [typelevel/cats#1567]
When you run our current `SplitLaw` for `Kleisli[Either[String, ?], ?, ?]` it fails the `splitInterchange` law.
@fthomas already mentioned in #232 that he wasn't sure that it was sound.

If you add the following to `KleisliTests` you can see it fail.
```scala
{
  type StringOr[A] = Either[String, A]
  
  // SI 2712 workarounds
  implicit val EqFAA     = implicitly[Eq[Kleisli[StringOr,Int,Int]]]
  implicit val EqFACB    = implicitly[Eq[Kleisli[StringOr,(Int, Int),Int]]]
  implicit val EqFACBC   = implicitly[Eq[Kleisli[StringOr,(Int, Int),(Int, Int)]]]
  implicit val EqFACDBCD = implicitly[Eq[Kleisli[StringOr,((Int, Int), Int),(Int, (Int, Int))]]]
  
  implicit val invariant = Kleisli.catsDataFunctorForKleisli[StringOr, Int]
  implicit val iso = CartesianTests.Isomorphisms.invariant[Kleisli[StringOr, Int, ?]]
  
  implicit val ArbFAB = implicitly[org.scalacheck.Arbitrary[Kleisli[StringOr,Int,Int]]]
  
  implicit val catsDataArrowForKleisli = Kleisli.catsDataSplitForKleisli[StringOr]
  checkAll("Kleisli[Either[String, ?], Int, Int]", SplitTests[Kleisli[StringOr, ?, ?]].split[Int, Int, Int, Int, Int, Int])
}
```

I looked at it with @diesalbla and the law only seems to hold for commutative arrows.

**Comment by diesalbla** [comment]
The law for the `split` operation from the `Split[F[_,_]` type-class, together with the laws enumerated for the `Arrow[_,_]` type class, has an structure which is not that of the conventional Arrows, as explained [here](https://wiki.haskell.org/Typeclassopedia#Arrow). Instead, it has the structure of a [Commutative Arrow](http://dl.acm.org/citation.cfm?doid=2976002.2976019), which is described in the referred article. That these laws are more restrictive means that some instances of the trait `Arrow[F[_,_]]` for some types `F` will hold only the normal _Arrow_ laws, but not the _Commutative Arrow_ ones. In particular, the instance constructor `implicit def catsDataArrowForKleisli[F[_]](implicit ev: Monad[F]): Arrow[Kleisli[F, ?, ?]]``` would not be a commutative arrow for `Kleisli`. 

##### Equivalence between Commutativity and `splitInterchange`.  
In `Haskell`-like notation, we write `***` for `split`, `>>>` for `compose`. The _commutativity_ law is that `first f >>> `second` g == `second` g >>> first f` for any `f,g`. The `splitInterchange` law is that `(f *** g) >>> (h >>> k)  == (f >>> h) *** (g >>> k)` for any `f,g,h,k`. We can expand this law as follows: 
```Haskell
(f *** g) >>> (h >>> k)  == (f >>> h) *** (g >>> k)  
first f >>> `second` g >>> first h >>> `second` k == first (f >>> h) >>> `second` (g >>> k)
first f >>> `second` g >>> first h >>> `second` k == first f >>> first h >>> `second` g >>> `second` k
```
By unfolding the definition of `***`, using the _functor_ law of `first` and `second`. We omit parenthesis because `>>>` is associative. From there, we see that _commutativity_ implies `splitChange`, because  it allows to swap the middle terms `second g` and `first h`. Also, `splitChange` imples _commutativity_, because if we apply it to the particular case `f=id, k=id`, we can remove the terms `first f` and `second k` to leave just the commutativity law.

**Comment by diesalbla** [comment]
Looking at the history, it seems that @non wrote the original definitions of the type-classes, but the `splitInterchange` law was later added by @fthomas in PR #232, but he comments there that he was not sure about its correctness. Moving forward, I can see two redesign options. 

The first option is to modify the inheritance hierarchy in the `cats.arrow` package: the trait `Arrow[F[_,_]]` should not inherit from `Split[F[_,_]]`, and we could create a class `CommutativeArrow[F[_,_]]` that would extend both `Split` and `Arrow`. The second option would be to simply remove the `splitInterchange` law. IMHO, the latter would just leave a hollow type class, so the former would be better.

**Comment by ceedubs** [comment]
@kailuowang should this be assigned to 1.0? It seems like we might be down a problematic trajectory on this one.

**Comment by kailuowang** [comment]
@ceedubs I agree. assigned.

**Comment by kailuowang** [comment]
I vote for option 1 proposed by @diesalbla in https://github.com/typelevel/cats/issues/1567#issuecomment-288121078, whose analysis in this issue is, IMO, great work.

**Comment by edmundnoble** [comment]
I'm good with option 1 as well.

**Comment by peterneyens** [comment]
`ArrowLaws` already has [a law covering the `split` method](https://github.com/typelevel/cats/blob/master/laws/src/main/scala/cats/laws/ArrowLaws.scala#L28-L29) and overrides [`split`](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/arrow/Arrow.scala#L23-L24).

We can make `split` a normal method of `Arrow` and add `CommutativeArrow` with the current `splitInterchange` law?

**Comment by fthomas** [comment]
Sorry for chiming in so late in the conversation. I think @diesalbla's and @peterneyens' analysis is spot-on. My justification for adding `splitInterchange` in #232  was based on this [proof](https://gist.github.com/fthomas/77a9a0a529f24e04ecc4) where I mistakenly assumed `second g >>> first f = first f >>> second g` for the arrow exchange law (which holds for conventional arrows). But as @diesalbla already noted, this is the commutativity law of commutative arrows.

**Comment by peterneyens** [comment]
Fixed in #1766.
